util/exercise/src/structs.rs

```diff
@@ -0,0 +1,159 @@
+//! `serde::Deserialize`.
+
+use crate::{self as exercise};
+use indexmap::IndexMap;
+use serde::{Deserialize, Serialize};
+
+macro_rules! pub_struct_serde_getters {
+    (
+        pub struct $name:ident {
+            $( $fname:ident : $ftype:ty ),* $(,)*
+        }
+    ) => {
+        #[derive(Serialize, Deserialize, Debug)]
+        pub struct $name {
+            $($fname : $ftype),*
+        }
+
+        impl $name {
+            $(
+            pub fn $fname(&self) -> &$ftype {
+                &self.$fname
+            }
+            )*
+        }
+    };
+}
+
+pub_struct_serde_getters! {
+    pub struct CanonicalData {
+        exercise: Exercise,
+        version: Version,
+        comments: Option<Comments>,
+        cases: TestGroup,
+    }
+}
+
+type Exercise = String;
+type Version = String;
+type Comments = Vec<String>;
+type TestGroup = Vec<LabeledTestItem>;
+
+#[derive(Serialize, Deserialize, Debug)]
+#[serde(untagged)]
+pub enum LabeledTestItem {
+    Single(LabeledTest),
+    Array(LabeledTestGroup),
+}
+
+pub_struct_serde_getters! {
+    pub struct LabeledTest {
+        description: Description,
+        optional: Option<Optional>,
+        comments: Option<Comments>,
+        property: Property,
+        input: Input,
+        expected: Expected,
+    }
+}
+
+impl LabeledTest {
+    pub fn generate_test_function(&self) -> String {
+        // Attributes.
+        let mut s = vec![String::from("#[test]"), String::from("#[ignore]")];
+
+        // Description and comments.
+        s.push(format!("/// {}", self.description()));
+        if let Some(comments) = self.comments() {
+            for line in comments {
+                s.push(format!("/// {}", line));
+            }
+        }
+
+        // Function signature (plus opening bracket).
+        s.push(format!(
+            "fn test_{}() {{",
+            exercise::format_exercise_description(self.description())
+        ));
+
+        // Unpack the inputs.
+        let mut inputs = Vec::new();
+        for (name, value) in self.input() {
+            inputs.push(name.clone());
+            s.push(format!("let {} = {};", name, value.to_literal()));
+        }
+        let inputs = inputs.join(", ");
+
+        // Output.
+        s.push(format!("let expected = {};", self.expected.to_literal()));
+
+        // Add call to the processing function (and closing brackets).
+        s.push(format!(
+            "process_{}_case({}, expected);\n}}\n",
+            exercise::format_exercise_property(self.property()),
+            inputs
+        ));
+
+        // Add newlines to pretty it up.
+        s.join("\n")
+    }
+}
+
+pub_struct_serde_getters! {
+    pub struct LabeledTestGroup {
+        description: Description,
+        optional: Option<Optional>,
+        comments: Option<Comments>,
+        cases: TestGroup,
+    }
+}
+
+type Description = String;
+type Optional = String;
+type Property = String;
+type Input = IndexMap<String, Data>;
+type Expected = Data;
```

Comment:
I was tempted to do so because it feels weird to split them, but, on the other hand, this version has the advantage of shadowing very closely the JSON schema (although the `impl LabeledTest` from lines 64 to 104 muddies things a bit).

This is not a great reason to split them though, as it is less readable this way. I'll see how it looks without the macro calls and the intervening `impl`, but I think it will be better to group them.

Reply:
I moved the `LabeledTest` implementation to after the structs and types. Having the types split up still feels a bit weird, but, combined with getting rid of the macro, it is easier to compare with the [JSON schema](https://github.com/exercism/problem-specifications/blob/master/canonical-schema.json). It made me notice that the `Expected` type might need some work to include an optional Error.

That being said, the structure is relatively simple, so grouping all the types shouldn't make maintenance/improvement very difficult.